### PR TITLE
[feature] allow changing maxPayload of open socket

### PR DESCRIPTION
**Pull request by mvduin**
This enables using a small maxPayload by default and then switching to a large value after the client has been authenticated in some way.

### CODE REVIEW

**Comment by lpinca** [comment]
Can you please elaborate on why this is useful? Isn't it sufficient to always use a large enough value?

**Comment by mvduin** [comment]
For the same reason maxPayload is useful in the first place: to help protect the server. Having a 100MB maxPayload means an attacker can easily consume that amount of the server's memory per connection established. Having a small maxPayload means an attacker would instead have to resort to flooding messages or connections, both of which can be dealt with by other means.

**Comment by lpinca** [comment]
Yes but why changing it dynamically? If you know your clients will say 1 MiB at max just always set it to 1 MiB no? What's the point of changing the value during the lifetime of connection?

**Comment by mvduin** [comment]
Because it may be difficult to evaluate how large it needs to be. We trust our clients however, and they have to authenticate themselves before they are allowed to send any "real" messages, so all that matters is that connections are prevented from sending large payloads until this authentication exchange has completed.

**Comment by lpinca** [comment]
Is there any reason to authenticate the clients after the WebSocket is opened? Can't it be done during the handshake? Or do need to allow both authenticated and unauthenticated connections?

**Comment by mvduin** [comment]
Using websocket messages for the authentication exchange is a lot more flexible, and it's too late to change anyway without breaking compatibility with existing clients.

Is there any reason to disallow modification of maxPayload after socket creation? It seems like a completely unnecessary limitation to me.

**Comment by lpinca** [comment]
I'm not super happy because it's not a standard attribute. It's a deviation from the WHATWG interface.

**Comment by mvduin** [comment]
But isn't that true for maxPayload in general already?

**Comment by lpinca** [comment]
Yes but it's not exposed `ws.maxPayload` :)

**Comment by mvduin** [comment]
I don't see how it's a good thing to hide an attribute that exists, has a functional effect, and whose initialization is already part of the API.

Also, it is mainly important for connections of a websocket server, which isn't covered by WHATWG at all anyway.

File: test/websocket.test.js [hunk 1]
```diff
@@ -164,6 +164,59 @@ describe('WebSocket', function() {
       });
     });
 
+    describe('`maxPayload`', function() {
+      it('defaults to 100MB upon "open"', function(done) {
```
Comment (lpinca) [inline test/websocket.test.js]:
```suggestion
      it('defaults to 100 MiB upon "open"', function(done) {
```

File: test/websocket.test.js [hunk 2]
```diff
@@ -164,6 +164,59 @@ describe('WebSocket', function() {
...
+          const ws = new WebSocket(`ws://localhost:${wss.address().port}`);
+
+          ws.onopen = () => {
+            assert.strictEqual(ws.maxPayload, 100 * 1024 * 1024);
+            wss.close(done);
+          };
+        });
+      });
+
+      it('can be overridden using options.maxPayload', function(done) {
```
Comment (lpinca) [inline test/websocket.test.js]:
```suggestion
      it('can be overridden using the `maxPayload` option', function(done) {
```

File: test/websocket.test.js [hunk 3]
```diff
@@ -164,6 +164,59 @@ describe('WebSocket', function() {
...
+        const maxPayload = 123;
+
```
Comment (lpinca) [inline test/websocket.test.js]:
```suggestion
```

File: test/websocket.test.js [hunk 4]
```diff
@@ -164,6 +164,59 @@ describe('WebSocket', function() {
       });
     });
 
+    describe('`maxPayload`', function() {
+      it('defaults to 100MB upon "open"', function(done) {
+        const wss = new WebSocket.Server({ port: 0 }, () => {
+          const ws = new WebSocket(`ws://localhost:${wss.address().port}`);
+
+          ws.onopen = () => {
```
Comment (lpinca) [inline test/websocket.test.js]:
```suggestion
          ws.on('open', () => {
```

File: test/websocket.test.js [hunk 5]
```diff
@@ -164,6 +164,59 @@ describe('WebSocket', function() {
...
+            assert.strictEqual(ws.maxPayload, 100 * 1024 * 1024);
+            wss.close(done);
+          };
```
Comment (lpinca) [inline test/websocket.test.js]:
```suggestion
          });
```

File: test/websocket.test.js [hunk 6]
```diff
@@ -164,6 +164,59 @@ describe('WebSocket', function() {
...
+
+      it('can be overridden using options.maxPayload', function(done) {
+        const maxPayload = 123;
+
+        const wss = new WebSocket.Server({ port: 0 }, () => {
+          const ws = new WebSocket(`ws://localhost:${wss.address().port}`, {
+            maxPayload
+          });
+
+          ws.onopen = () => {
```
Comment (lpinca) [inline test/websocket.test.js]:
```suggestion
          ws.on('open', () => {
```

File: test/websocket.test.js [hunk 7]
```diff
@@ -164,6 +164,59 @@ describe('WebSocket', function() {
...
+            assert.strictEqual(ws.maxPayload, maxPayload);
+            wss.close(done);
+          };
```
Comment (lpinca) [inline test/websocket.test.js]:
```suggestion
          });
```

File: test/websocket.test.js [hunk 8]
```diff
@@ -164,6 +164,59 @@ describe('WebSocket', function() {
...
+        });
+      });
+
+      it('can be modified after "open"', function(done) {
+        const maxPayload = 123;
+
```
Comment (lpinca) [inline test/websocket.test.js]:
```suggestion
```

File: test/websocket.test.js [hunk 9]
```diff
@@ -164,6 +164,59 @@ describe('WebSocket', function() {
...
+          });
+
+          ws.onopen = () => {
+            assert.strictEqual(ws.maxPayload, maxPayload);
+            wss.close(done);
+          };
+        });
+      });
+
+      it('can be modified after "open"', function(done) {
```
Comment (lpinca) [inline test/websocket.test.js]:
```suggestion
      it("can be modified after the 'open' event is emitted", function(done) {
```

File: test/websocket.test.js [hunk 10]
```diff
@@ -164,6 +164,59 @@ describe('WebSocket', function() {
...
+
+        const wss = new WebSocket.Server(
+          {
+            perMessageDeflate: true,
+            port: 0
+          },
+          () => {
+            const ws = new WebSocket(`ws://localhost:${wss.address().port}`);
+
+            ws.onopen = () => {
```
Comment (lpinca) [inline test/websocket.test.js]:
```suggestion
            ws.on('open', () => {
```

File: test/websocket.test.js [hunk 11]
```diff
@@ -164,6 +164,59 @@ describe('WebSocket', function() {
...
+              ws.maxPayload = maxPayload;
+              assert.strictEqual(ws.maxPayload, maxPayload);
+              assert.strictEqual(ws._receiver._maxPayload, maxPayload);
+              assert.strictEqual(
+                ws._receiver._extensions['permessage-deflate']._maxPayload,
+                maxPayload
+              );
+              wss.close(done);
+            };
```
Comment (lpinca) [inline test/websocket.test.js]:
```suggestion
            });
```

**Comment by lpinca** [comment]
Please add documentation.

**Comment by lpinca** [comment]
> I don't see how it's a good thing to hide an attribute that exists, has a functional effect, and whose initialization is already part of the API.

Private attributes/methods exist in many languages for a reason. The with JS is that they are not that private.

**Comment by mvduin** [comment]
> Nits

Will fix. I didn't like the onopen either, but I just copy-pasted it from the existing testcode below the one I inserted. Whenever I edit code that's not mine I default to using whatever style the code in the immediate surroundings uses, but I see now it just happened to be one of the small minority of test cases that still use onopen.

> Please also add documentation.

Will fix!

> Private attributes/methods exist in many languages for a reason.

I'm not disputing that, but maxPayload is already user-provided, and I don't see any good reason to keep it private after initialization.

**Comment by lpinca** [comment]
@websockets/admin @3rd-Eden what do you think?

**Comment by mvduin** [comment]
My previous version didn't allow maxPayload to be get/set on client sockets while connecting (i.e. before _receiver is set) because the setting was being held in a closure at that time. While I don't care about the ability to modify on client sockets at all, let alone when they're connecting, this exception just seemed too weird and unsightly to leave it be, especially as I tried to document it.

I've fixed this by keeping the setting on the WebSocket itself and syncing it to _receiver (and deflate extension, if applicable), similar to how binaryType is handled.

File: lib/websocket.js [hunk 12]
```diff
@@ -457,6 +474,7 @@ function initAsClient(address, protocols, options) {
   }
 
   this._isServer = false;
+  this.maxPayload = options.maxPayload;
```
Comment (lpinca) [inline lib/websocket.js]:
```suggestion
  this._maxPayload = options.maxPayload;
```

File: lib/websocket-server.js [hunk 13]
```diff
@@ -192,8 +192,7 @@ class WebSocketServer extends EventEmitter {
     if (this.options.perMessageDeflate) {
       const perMessageDeflate = new PerMessageDeflate(
         this.options.perMessageDeflate,
-        true,
-        this.options.maxPayload
```
Comment (lpinca) [inline lib/websocket-server.js]: Unnecessary change.

File: lib/websocket.js [hunk 14]
```diff
@@ -47,6 +47,7 @@ class WebSocket extends EventEmitter {
     this._closeCode = 1006;
     this._extensions = {};
     this._isServer = true;
+    this._maxPayload = 0;
```
Comment (lpinca) [inline lib/websocket.js]: I'd use the same default value of the `maxPayload` option.

**Comment by lpinca** [comment]
@mvduin another thing I was thinking about is that changing the value dynamically may have surprising effects if that is done while the payload of a frame is being decompressed or in the middle of a fragmented message. I think it's like giving people a loaded gun to shoot themselves in the foot with, so if anything changing it dynamically should be discouraged.

**Comment by lpinca** [comment]
Closing as per https://github.com/websockets/ws/pull/1473#issuecomment-441112214 and https://github.com/websockets/ws/pull/1473#issuecomment-441778987.

I think that changing the value of `maxPayload` dynamically is not something that ws should officially support/endorse. It's still possible to achieve this by using the private `_maxPayload` property on the `Receiver` and `PerMessageDeflate` instances if really needed.

I'm happy to reopen this if someone can make me change my mind.
